From b43eaad61f72a571a3360c04bbb36fc05740bf86 Mon Sep 17 00:00:00 2001
From: Simon Shields <simon@lineageos.org>
Date: Tue, 29 Aug 2017 23:03:29 +1000
Subject: [PATCH 03/10] edify: bring back SetPermissionsRecursive

* we still need this, even with block-based OTAs.

This partially reverts commit f388104eaacd05cfa075d6478369e1d0df5ddbf3

includes the following commit:

commit d810b09316cb1abc55b9d44ac008a4245ba642fe
Author: LuK1337 <priv.luk@gmail.com>
Date:   Fri Sep 1 00:12:29 2017 +0200

    edify: Don't append capabilities if path starts with /tmp

    * tmpfs does not support xattrs.

    Change-Id: I90ce6455e3da91d3018f2f2a427e398281e654ab

Change-Id: I4943e2e89ee5c810a63746c570dc5e31e95b8c53
---
 core/Makefile                         |  1 +
 tools/releasetools/edify_generator.py | 19 +++++++++++++++++++
 2 files changed, 20 insertions(+)

diff --git a/core/Makefile b/core/Makefile
index ec765710a..73af7f803 100644
--- a/core/Makefile
+++ b/core/Makefile
@@ -2798,6 +2798,7 @@ ifdef PRODUCT_EXTRA_RECOVERY_KEYS
 endif
 	$(hide) echo 'mkbootimg_args=$(BOARD_MKBOOTIMG_ARGS)' >> $(zip_root)/META/misc_info.txt
 	$(hide) echo 'mkbootimg_version_args=$(INTERNAL_MKBOOTIMG_VERSION_ARGS)' >> $(zip_root)/META/misc_info.txt
+	$(hide) echo "use_set_metadata=1" >> $(zip_root)/META/misc_info.txt
 	$(hide) echo "multistage_support=1" >> $(zip_root)/META/misc_info.txt
 	$(hide) echo "blockimgdiff_versions=3,4" >> $(zip_root)/META/misc_info.txt
 ifneq ($(OEM_THUMBPRINT_PROPERTIES),)
diff --git a/tools/releasetools/edify_generator.py b/tools/releasetools/edify_generator.py
index 16e2fc9c4..8b5b43969 100644
--- a/tools/releasetools/edify_generator.py
+++ b/tools/releasetools/edify_generator.py
@@ -301,6 +301,25 @@ class EdifyGenerator(object):
     cmd = "".join(cmd)
     self.script.append(self.WordWrap(cmd))
 
+  def SetPermissionsRecursive(self, fn, uid, gid, dmode, fmode, selabel,
+                              capabilities):
+    """Recursively set path ownership and permissions."""
+    if not self.info.get("use_set_metadata", False):
+      self.script.append('set_perm_recursive(%d, %d, 0%o, 0%o, "%s");'
+                         % (uid, gid, dmode, fmode, fn))
+    else:
+      if capabilities is None:
+        capabilities = "0x0"
+      cmd = 'set_metadata_recursive("%s", "uid", %d, "gid", %d, ' \
+          '"dmode", 0%o, "fmode", 0%o' \
+          % (fn, uid, gid, dmode, fmode)
+      if not fn.startswith("/tmp"):
+        cmd += ', "capabilities", "%s"' % capabilities
+      if selabel is not None:
+        cmd += ', "selabel", "%s"' % selabel
+      cmd += ');'
+      self.script.append(cmd)
+
   def WriteRawImage(self, mount_point, fn, mapfn=None):
     """Write the given package file into the partition for the given
     mount point."""
-- 
2.17.1

